algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+begin
+  intro hf,
+  cases hf,
+  constructor,
+  intros S _,
+  cases classical.em (∃ x, x ≠ (0:α) ∧ x ∈ S),
+  right,
+  cases h with x hx,
+  apply is_ideal.eq_univ_of_contains_unit S,
+  exact ⟨x, hx.2, hf_h hx.1⟩,
+  left,
+  apply set.ext,
+  intro x,
+  split,
+  intro hx,
+  apply mem_singleton_of_eq,
+  apply @of_not_not _ (classical.prop_decidable _),
+  intro hnx,
+  exact h ⟨x, hnx, hx⟩,
+  intro hx,
+  unfold zero_ideal at hx,
+  rw mem_singleton_iff at hx,
+  rw hx,
+  exact is_ideal.zero_mem S
+end
+
+theorem ideal_eq_zero_or_univ.to_hom_inj : ideal_eq_zero_or_univ α → hom_inj α :=
+λ ⟨h⟩, ⟨λ β _ hne f _ x y hxy, begin
+  cases h (is_hom.ker f) with hkz hku,
+  { have : x - y ∈ is_hom.ker f,
+    { by simp [is_hom.ker, is_hom.map_add f, is_hom.map_neg f, hxy] },
+    simpa [hkz, add_neg_eq_zero] using this },
+  { have : (1:α) ∈ is_hom.ker f,
+    { by rw hku; trivial },
+    simp [is_hom.map_one f] at this,
+    cc }
+end⟩
+
+theorem hom_inj.to_is_field : hom_inj α → is_field α :=
+begin
+  intro h,
+  cases h,
+  constructor,
+  intros x hx,
+  specialize h (α / principal_ideal x),
+  cases classical.em ((0 : α / principal_ideal x) = 1) with h1 h1,
+  have := @quotient.exact _ (is_ideal.setoid $ principal_ideal x) _ _ h1,
+  change (0:α) - 1 ∈ principal_ideal x at this,
+  have : (1:α) ∈ principal_ideal x := calc
+    (1:α) = (0 - 1) * (-1) : by norm_num
+      ... ∈ principal_ideal x : is_ideal.mul_mem this,
+  exact this,
+  specialize @h h1 _ (is_ideal.to_quotient $ principal_ideal x) x 0,
+  exfalso,
+  apply hx,
+  apply h,
+  apply is_ideal.coset_zero_of_mem,
+  exact mem_principal_ideal x,
+  exact zero_ne_one
+end
+
+end prop_1_2
+
+-- Proposition 1.2 end
+
+
+section prime_ideals_and_maximal_ideals
+
+variables {α : Type u} [comm_ring α] (S : set α) [is_ideal α S]
+
+class is_prime_ideal : Prop :=
+(ne_univ_ideal : S ≠ univ_ideal α)
+(mem_or_mem_of_mul_mem : ∀ {x y : α}, x * y ∈ S → x ∈ S ∨ y ∈ S)
+
+class is_maximal_ideal : Prop :=
+(ne_univ_ideal : S ≠ univ_ideal α)
+(no_between : ∀ (T : set α) [is_ideal α T], S ⊆ T → T = S ∨ T = univ_ideal α)
+
+variable α
+
+class is_integral_domain : Prop :=
+(zero_ne_one : 0 ≠ (1:α))
+(zero_or_zero_of_mul_zero : ∀ a b : α, a * b = 0 → a = 0 ∨ b = 0)
+
+variable {α}
+
+theorem prime_iff_quotient_integral_domain : is_prime_ideal S ↔ is_integral_domain (α/S) :=
+⟨λ ⟨hne, hmul⟩,
+   ⟨λ hzo, hne $ is_ideal.univ_of_one_mem S 
+      (by simpa [is_ideal.mem_iff_coset_zero S] using hzo.symm),
+    λ x y, let ⟨m, hm⟩ := is_ideal.coset_rep x,
+               ⟨n, hn⟩ := is_ideal.coset_rep y in
+      by rw [← hm, ← hn]; simpa [is_ideal.coset_zero_iff_mem S] using hmul⟩,
+ λ ⟨hne, hmul⟩,
+   ⟨λ hsu, hne $ eq.symm $ is_ideal.coset_zero_of_mem S 1 $ by rw hsu; trivial,
+    λ x y, by simpa [is_ideal.mem_iff_coset_zero S] using hmul ⟦x⟧ ⟦y⟧⟩⟩
+
+theorem maximal_iff_quotient_field : is_maximal_ideal S ↔ is_field (α/S) :=
+begin
+  split,
+  intro h,
+  cases h,
```

Comment:
After split, intro h should be indented right? Even better might be { intro h... I guess.